static mut: Add `SingleThreadValue` which checks the running thread

bradjc opened the pull request:
### Pull Request Overview

Any holder of `SingleThreadValue` can only access the contained value if the execution is on the same thread that initialized the `SingleThreadValue`.

I had to create a new trait `ChipThreadId` because `Chip` cannot be held in a `dyn` reference.


### Testing Strategy

This pull request was tested by...


### TODO or Help Wanted

This pull request still needs...


### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

alevy commented on chips/nrf52/src/chip.rs:
```diff
@@ -150,3 +150,9 @@ impl<'a, I: InterruptService + 'a> kernel::platform::chip::Chip for NRF52<'a, I>
+
+impl<'a, I: InterruptService + 'a> kernel::platform::chip::ChipThreadId for NRF52<'a, I> {
+    fn running_thread_id(&self) -> usize {
+        1
```
Except this would actually need to be dynamically determined, right? E.g. in this case by looking at the IPSR register

bradjc replied:
Yes.

ZhekaS replied:
How will it do on architectures that do not have hard indication of being "inside" an ISR? RiscV I believe is one. Perhaps something similar to [CriticalSection](https://github.com/rust-embedded/critical-section) would be more versatile?

bradjc replied:
We could fall back to some sort of manual tracking, e.g., all ISRs somehow mark they are executing and `running_thread_id()` uses that. That is of course error prone and not the easiest to enforce, but, it doesn't seem to me to be worse than just not using the notion of a thread ID and just falling back to trusting the programmer correctness anyway for this SingleThreadValue type.

lschuermann commented on kernel/src/utilities/single_thread_value.rs:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    pub fn set_chip(&self, chip: &'static dyn crate::platform::chip::ChipThreadId) {
+        self.chip.set(chip);
+        self.thread_id.set(chip.running_thread_id());
+    }
```
This would have to check whether its already initialized and, if so, refuse this operation.

lschuermann replied:
It would also have to be an `unsafe` method, because you assert that this is the currently running chip instance, and not some other instance you constructed.

lschuermann replied:
Yes, they could simply instantiate two `Chip` instances.

bradjc replied:
But how? They would have to implement the `unsafe` trait, which would guarantee that the new `chip` still upholds the specified invariants. I'm thinking that marking the trait unsafe means that all implementations are by definition valid (or, we don't care that they are broken).

lschuermann replied:
What makes me uncomfortable is that that, right now, there's nothing stopping an implementation of `running_thread_id` to rely on some `Chip`-object specific context (such as a marker to indicate whether we're currently in a trap handler or not).

It is true that an `unsafe trait` forces the implementer of a type implementing `ChipThreadId` to assert that this implementation is correct. However, an implementation of this trait can be correct under the assumption that you're always calling it on the "real" chip object that's being passed to the kernel.

I think there's two solutions:

1. We can simply force the `running_thread_id` method to not have an object context. This way, there's no "wrong object" the user could be possibly passing in, and no way for `running_thread_id`'s behavior to be specific to any object.

   We could do so through a trait method that doesn't take `&self`. This trait would still be an `unsafe trait`, and your exact reasoning applies:

   > I'm thinking that marking the trait unsafe means that all implementations are by definition valid [...]

   Thus, with this, `set_chip` could be safe.

2. Alternatively, we'd mark this method as unsafe, because that forces the *user* to attest that they are passing in the actual `ChipThreadId` instance, and not some other instance that they've created but is not the actual `Chip` being used by other kernel infrastructure, etc.

   One case where this becomes a real concern is in the multi-core setting. In that setting, you'd conceivably hold two `Chip` instances, concurrently (one for each core). Then, the user, when initializing this struct, must guarantee that they are passing in the correct instance for the core they're currently on.

3. We can simply require implementers of this trait to not rely on any object state in their implementation of `running_thread_id`. That's pretty much a less elegant version of (1), which can't be enforced through the type system.

bradjc replied:
> It is true that an `unsafe trait` forces the implementer of a type implementing `ChipThreadId` to assert that this implementation is correct. However, an implementation of this trait can be correct under the assumption that you're always calling it on the "real" chip object that's also being passed to the kernel.

Seems to me that the easy solution is to say this assumption is invalid and that such an implementation is in fact not correct, thus violating the guarantees of implementing the trait. If calling `running_thread_id()` does not in fact return the running thread's ID, that seems like an implementation problem, not our problem.

bradjc replied:
Seems fine now, let's make sure we can implement for our three+ `arch`s

lschuermann commented on chips/nrf52/src/chip.rs:
```diff
@@ -150,3 +150,9 @@ impl<'a, I: InterruptService + 'a> kernel::platform::chip::Chip for NRF52<'a, I>
     }
 }
+
+impl<'a, I: InterruptService + 'a> kernel::platform::chip::ChipThreadId for NRF52<'a, I> {
```
This would have to be an `unsafe trait`, because we rely on implementations of this trait to be correct.

lschuermann reviewed (commented):
This is close to what we've discussed offline.

I don't think it's necessary to hold an actual dynamic-dispatch, fat-pointer reference to the `Chip`. We can avoid dynamic dispatch by just taking a function pointer that determines the currently running thread. I don't believe that this would actually need access to any `Chip`-instance variables to do so (presumably there's a CPU register to query on Cortex-M? For RISC-V, we can look at `mstatus` and `mhartid`).

bradjc commented:
Hopefully I'm reading the room right and added a bunch of comments. I also included my summary of how this type was designed as a `//` comment.

I also removed the `unsafe` from the constructor. The comment about safety no longer applies, and I don't think the new() function needs to be unsafe.

brghena commented on kernel/src/utilities/single_thread_value.rs:
```diff
@@ -0,0 +1,186 @@
+/// # Guaranteeing Single-Thread Access
+///
+/// [`SingleThreadValue`] is safe because it guarantees that the value is only
+/// ever accessed from the same thread that created it. To do this,
```
```suggestion
/// ever accessed from the same thread that called `set_chip`. To do this,
```

bradjc replied:
Absolutely, and that "thread" has created and now owns this value. Hm, maybe we should rename set_chip to `create`.

brghena commented on kernel/src/utilities/single_thread_value.rs:
```diff
@@ -0,0 +1,186 @@
+    /// This stores the method that can identify the currently executing thread.
+    /// This method is used to determine if an attempted access is permitted or
+    /// not.
+    pub fn set_chip<C: crate::platform::chip::ChipThreadId>(&self) {
```
I'm not sure why this is called `set_chip`? It's not _just_ setting the `thread_id`, but it's not setting the chip either. Maybe something like `initialize_thread()`?

bradjc replied:
I agree, the naming in this PR is generally terrible. However, I'm wary of a name discussion not converging.

bradjc replied:
The irony in that a type that's taken 1.5 years to create deserves careful naming while a type that's taken this long just needs to get created.

brghena commented on kernel/src/utilities/single_thread_value.rs:
```diff
@@ -0,0 +1,186 @@
+        }
+    }
+
+    /// Assign the [`ChipThreadId`] implementation.
+    ///
+    /// This stores the method that can identify the currently executing thread.
+    /// This method is used to determine if an attempted access is permitted or
+    /// not.
```
This also sets the `thread_id` to whichever thread ran this function, which seems important to document. You could, for instance, make something only accessible from interrupts if you called `set_chip()` from an interrupt context initially.

brghena reviewed (commented):
This makes sense to me and seems good.

I really appreciate that there are far more comments (especially about safety) than there is code.

I left some comments on things as I read through.

lschuermann commented on kernel/src/utilities/single_thread_value.rs:
```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    /// Assign the [`ChipThreadId`] implementation.
+    ///
+    /// This stores the method that can identify the currently executing thread.
+    /// This method is used to determine if an attempted access is permitted or
+    /// not.
+    pub fn set_chip<C: crate::platform::chip::ChipThreadId>(&self) {
```
I hate to say this, but I think this current implementation is still unsound... ☹️ 

This is because we can now have multiple, concurrent calls to `set_chip`, `which`, or `with_valid` that are accessing shared state (the `thread_id_and_fn` field) which is itself not thread safe. And the only reason why this passes the compiler is that we manually implement `Sync` for this type -- promising that it is thread safe, when it really isn't.

There's a simple solution here:

```suggestion
    /// Assign the [`ChipThreadId`] implementation.
    ///
    /// This stores the method that can identify the currently executing thread.
    /// This method is used to determine if an attempted access is permitted or
    /// not.
    ///
    /// # Safety
    ///
    /// Callers of this function must ensure that there are no concurrent calls
    /// to any functions that access (read or write) the internal fields of this
    /// struct storing the thread ID this type is initialized for, and the
    /// function pointer used to obtain this thread ID. This includes the
    /// `set_chip` function itself, as well as all other methods that check the
    /// thread ID, such as `with` and `with_valid`.
    ///
    /// A simple way to maintain this guarantee is to only ever call this
    /// function once, before any other uses (reads / writes) of this type.
    pub unsafe fn set_chip<C: crate::platform::chip::ChipThreadId>(&self) {
```

By making the method _writing_ to this shared state unsafe, we can shift the burden of ensuring that there are no concurrent, conflicting accesses to the user. Simply write it once at the start, and only then are you allowed to interact with it otherwise.

However, even that might not be 100% sufficient, because the compiler is free to assume that there is no happens-before relation between performing this call to `set_chip` and any other read or write to this data structure from another thread. We'd also need to insert appropriate [`fence`s](https://doc.rust-lang.org/stable/core/sync/atomic/fn.fence.html) in `set_chip`, `with`, and `with_valid`.

If this `unsafe` really is unbearable for us then I have an idea of how we could work around these issues with some assembly. I'll keep experimenting more over the next few days.

brghena replied:
Ugh. This turns into a readers/writer problem, huh?

lschuermann replied:
Yes, essentially. At least it looks like there's an ugly but portable solution around the _compiler fence does not work for non-atomic writes_ issue: it looks like an empty `asm!` block, which should be portable across all platforms, **without** `nomem` does the trick: https://github.com/taiki-e/portable-atomic/blob/7a0b15b867790ad384ae0e8edaddc7d8b4e75800/src/imp/msp430.rs#L48-L68

But it's still really tricky to reason through, and ideally we'd just have our own "fake-atomic" primitives in the kernel for single-core chips, as implemented by [`portable-atomic`](https://github.com/taiki-e/portable-atomic/blob/7a0b15b867790ad384ae0e8edaddc7d8b4e75800/src/imp/msp430.rs#L48-L68). But that crate is *huge* and full of macros, so we'd want to vendor our own. And finally, because we can't have a circular dependency of `arch -> kernel -> arch`, we'd probably need to introduce arch-specific code into the kernel. Yuck.

lschuermann replied:
Here's what that would look like with these DIY "compiler fences": https://github.com/tock/tock/commit/785c3264da35ed305c563e85fdc1968a662a7651

I don't know whether that's actually sufficient for all platforms (esp. ones that do instruction reordering in the CPU and expose a weird memory model to the user), but it should at least cause `rustc` / LLVM to flush out all writes before hitting one of these `asm!()` blocks and avoid making assumptions about the contents of any `UnsafeCell` post an `asm!()` block.

bradjc replied:
We still cannot use atomics in the kernel?

ZhekaS replied:
May I ask why atomics can't be used? I know Rust RISC-V targets without `A` extension used not to have atomic types support, but I believe it is using the LLVM's `+forced-atomics`  now and the primitive atomic types are supported. Are there other cases which restrict the usage?

lschuermann replied:
Yes, if we can make use of `+forced-atomics`, and the necessary `__sync_*` implementations are available, then that'd be great! It would mean that the kernel crate wouldn't work for any target that doesn't support these primitive atomic types. But it'd most certainly be the cleanest solution for now.

I know that this _did not_ used to work a while back (where we had similar issues around deferred call, that we've punted on because it's only accessed by one thread in practice), so that sent me down this rabbit hole.

lschuermann replied:
#4551 is the result of our lengthy offline discussion, which uses `AtomicUsize::{load, store}` (as provided by `+forced-atomics` for some our our platforms that don't have "rich" support for atomics in hardware, like `rv32i`) to work around the fencing and happens-before issue.

Where available, it also uses `compare_exchange` to prevent issues due to concurrent initialization. Where that's not available, we instead require the developer to uphold the invariant that there are no concurrent calls to `bind_to_thread` (né `set_chip`) -- we can't really do much better in practice.

lschuermann reviewed (commented):
I pushed two commits, one to consolidate the two `OptionalCells` into one and also make use of niche filling (because Rust's function references can never be `NULL`, we'll save a word of memory), and to fix a clippy lint.

I also spotted a soundness bug in the current implementation. I'm decently sure we can find a solution for it, but it shows how careful we need to be for this to actually be sound. I have two (hopefully) sound solutions in mind, that I can draft up in the next few days.

lschuermann commented on kernel/src/utilities/single_thread_value.rs:
```diff
@@ -0,0 +1,182 @@
+// generally.
+pub struct SingleThreadValue<T> {
+    value: T,
+    thread_id_and_fn: OptionalCell<(fn() -> usize, usize)>,
```
It's also one less check per operation. We've always loaded the contents from both `OptionalCell`s in a nested closure, which is unnecessary.
